modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java

```diff
@@ -660,4 +664,32 @@ private static TypeSpec convertToTypeSpec(RelDataType type) {
             throw new IllegalArgumentException("Unexpected type: " + type);
         }
     }
+
+    /** Check limitation for character types and throws exception if row contains character sequence more than type defined. */
+    public static <RowT> void validateCharactersOverflow(RelDataType rowType, RowT row, RowHandler<RowT> rowHandler) {
+        int colCount = rowType.getFieldList().size();
+
+        for (int i = 0; i < colCount; ++i) {
+            RelDataType colType = rowType.getFieldList().get(i).getType();
+            if (CHAR_TYPES.contains(colType.getSqlTypeName())) {
+                Object data = rowHandler.get(i, row);
+
+                if (data == null) {
+                    continue;
+                }
+
+                assert data instanceof String;
+
+                String str = (String) data;
+
+                int colPrecision = colType.getPrecision();
+
+                assert colPrecision != RelDataType.PRECISION_NOT_SPECIFIED;
+
+                if (str.stripTrailing().length() > colPrecision) {
```

Comment:
so, we use stripped string just to check length and then throw it away... this doesn't look right. We should not put to storage rows with string values whose length violates specified constraint.

```
CREATE TABLE t (id INT PRIMARY KEY, val VARCHAR(3));
INSERT INTO t VALUES (1, ' 1  ');
SELECT length(val) FROM t; -- returns 4
```

the query above returns 4, but I would expect 3 to be returned in that case